dungeon/src/contrib/systems/FogSystem.java

```diff
@@ -0,0 +1,359 @@
+package contrib.systems;
+
+import contrib.utils.EntityUtils;
+import core.Entity;
+import core.Game;
+import core.System;
+import core.components.DrawComponent;
+import core.components.PositionComponent;
+import core.level.Tile;
+import core.level.utils.LevelElement;
+import core.level.utils.LevelUtils;
+import core.utils.Point;
+import core.utils.components.MissingComponentException;
+import java.util.*;
+
+/**
+ * The FogOfWarSystem class is responsible for controlling the fog of war in the game.
+ *
+ * <p>The fog of war is a game mechanic where areas of the game world that are not in the player's
+ * line of sight are obscured. This class maintains a set of tiles that are currently darkened (not
+ * visible to the player) and a list of entities that are hidden. It also keeps track of the last
+ * known position of the hero (player character) and whether the fog of war system is currently
+ * active.
+ */
+public class FogSystem extends System {
+  private static final int DISTANCE_TRANSITION_SIZE = 2; // size of distance transition (in tiles)
+  private static final int HIDE_ENTITY_THRESHOLD =
+      0xFFFFFF99; // tint color threshold for hiding entities
+  private static final int[][] mult = { // needed for casting light
+    {1, 0, 0, -1}, {0, 1, -1, 0}, {0, -1, -1, 0}, {-1, 0, 0, -1},
+    {-1, 0, 0, 1}, {0, -1, 1, 0}, {0, 1, 1, 0}, {1, 0, 0, 1}
+  };
+  private static final float TINT_COLOR_WALL_DISTANCE_SCALE =
+      1.5f; // scale factor for behind wall distance fog
+  private static final float TINT_COLOR_DISTANCE_SCALE = .5f; // scale factor for distance fog
+
+  /** The view distance (range for tiles that are fully visible). */
+  private static int currentViewDistance = 7;
+
+  /** The maximum view distance (all tiles to consider for calculation). */
+  private static final int MAX_VIEW_DISTANCE = 25;
+
+  private final Map<Tile, Integer> darkenedTiles = new HashMap<>();
+  private final List<Entity> hiddenEntities = new ArrayList<>();
+  private boolean active = true;
+
+  /**
+   * Resets the FogOfWarSystem.
+   *
+   * <p>This method clears the sets of darkened tiles and hidden entities.
+   *
+   * @param revert If true, the FogOfWarSystem will also revert to its initial state.
+   * @see #reset()
+   * @see #revert()
+   */
+  public void reset(boolean revert) {
+    darkenedTiles.clear();
+    hiddenEntities.clear();
+    if (revert) {
+      revert();
+    }
+  }
+
+  /**
+   * Resets the FogOfWarSystem to its initial state.
+   *
+   * <p>This method clears the sets of darkened tiles and hidden entities, and resets the last known
+   * hero position. The last known hero position is set to the current hero's position if a hero
+   * exists, otherwise it is set to (0,0). This method also reverts the FogOfWarSystem to its
+   *
+   * @see #revert()
+   */
+  public void reset() {
+    reset(true);
+  }
+
+  /** Reverts the FogOfWarSystem. This reveals all darkened tiles and hidden entities. */
+  public void revert() {
```

Comment:
Die Namensgebung ist noch nicht ganz glücklich. Diese Funktion schaltet das FoW quasi wirkungslos, aber nicht ab, oder? Es wird nichts im zurückgesetzt - nur sichtbar gemacht?

Reply:
Na doch, der Fog wird halt zurückgesetzt. 
Es schaltet nichts ab und macht nix wirkungslos. 
Es ist ja ein System, im nächsten `execute`/Frame wird "neuer Fog" erzeugt.

Reply:
eben. also irgendwie kein "revert" - das würde ja dinge dauerhaft zurückändern (for a given value of "dauerhaft"). ich finde den begriff aus dem hinteren teil des kommentars besser: `revealAll` oder so.

Reply:
habs mal umbenannt, ist mir zu egal um in 1..2 Monaten nochmal drüber zu reden :D